src/DropdownBase.js

```diff
@@ -0,0 +1,307 @@
+import React, { cloneElement } from 'react';
+import keycode from 'keycode';
+import DropdownToggle from './DropdownToggle';
+import DropdownMenu from './DropdownMenu';
+import MenuItem from './MenuItem';
+import CustomPropTypes from './utils/CustomPropTypes';
+import createChainedFunction from './utils/createChainedFunction';
+
+import find from 'lodash/collection/find';
+
+const TOGGLE_REF = 'toggle-btn';
+
+export default class DropdownBase extends React.Component {
+  constructor(props, Toggle=DropdownToggle) {
+    super(props);
+
+    this.Toggle = Toggle;
+
+    this.toggleOpen = this.toggleOpen.bind(this);
+    this.handleClick = this.handleClick.bind(this);
+    this.handleKeyDown = this.handleKeyDown.bind(this);
+    this.handleRequestClose = this.handleRequestClose.bind(this);
+    this.handleSelect = this.handleSelect.bind(this);
+    this.extractChildren = this.extractChildren.bind(this);
+
+    this.refineMenu = this.refineMenu.bind(this);
+    this.refineToggle = this.refineToggle.bind(this);
+
+    this.childExtractors = [{
+      key: 'toggle',
+      matches: child => child.type === this.Toggle,
+      refine: this.refineToggle
+    }, {
+      key: 'menu',
+      matches: child => child.type === DropdownMenu || child.type !== MenuItem,
```

Comment:
I am always a bit wary of type checks like this...I am not sure why someone might have multiple instances of RB but this would break if they did, perhaps that's correct...it does limit the ability to create custom menu components though. nit picky concerns i guess.